packages/blocks/src/types.ts

```diff
@@ -0,0 +1,976 @@
+				  }
+				| {
+						type: 'object';
+						default?: object | undefined;
+				  }
+				| {
+						type: 'boolean';
+						default?: boolean | undefined;
+				  }
+				| {
+						type: 'number';
+						default?: number | undefined;
+				  }
+				| {
+						type: 'string';
+						default?: string | undefined;
+				  }
+		   ) )
+		| 'array'
+		| 'object'
+		| 'boolean'
+		| 'number'
+		| 'string';
+}
+
+export type WPBlockAttribute< T > =
+	| AttributeSource.Attribute
+	| AttributeSource.Children
+	| AttributeSource.HTML
+	| AttributeSource.Meta
+	| AttributeSource.Query< T >
+	| AttributeSource.Text
+	| AttributeSource.None
+	| { [ key: string ]: any }; // Permit untyped/undocumented property access.
+
+//
+// Transforms
+// ----------------------------------------------------------------------------
+
+export type TransformRawSchema = {
+	[ k in keyof HTMLElementTagNameMap | '#text' ]?: {
+		attributes?: string[] | undefined;
+		require?: Array< keyof HTMLElementTagNameMap > | undefined;
+		classes?: Array< string | RegExp > | undefined;
+		children?: TransformRawSchema | undefined;
+		[ key: string ]: any; // Permit untyped/undocumented property access.
+	};
+};
+
+export interface TransformBlock< T extends Record< string, any > > {
+	type: 'block';
+	priority?: number | undefined;
+	blocks: string[];
+	isMatch?( attributes: T, block: string | string[] ): boolean;
+	isMultiBlock?: boolean | undefined;
+	transform( attributes: T ): WPBlock< Partial< T > >;
+	[ key: string ]: any; // Permit untyped/undocumented property access.
+}
+
+export interface TransformEnter< T extends Record< string, any > > {
+	type: 'enter';
+	priority?: number | undefined;
+	regExp: RegExp;
+	transform(): WPBlock< Partial< T > >;
+	[ key: string ]: any; // Permit untyped/undocumented property access.
+}
+
+export interface TransformFiles< T extends Record< string, any > > {
+	type: 'files';
+	priority?: number | undefined;
+	isMatch?( files: FileList ): boolean;
+	transform(
+		files: FileList,
+		onChange?: ( id: string, attrs: T ) => void
+	): WPBlock< Partial< T > >;
+	[ key: string ]: any; // Permit untyped/undocumented property access.
+}
+
+export interface TransformPrefix< T extends Record< string, any > > {
+	type: 'prefix';
+	priority?: number | undefined;
+	prefix: string;
+	transform( content: string ): WPBlock< Partial< T > >;
+	[ key: string ]: any; // Permit untyped/undocumented property access.
+}
+
+export interface TransformRaw< T extends Record< string, any > > {
+	type: 'raw';
+	priority?: number | undefined;
+	/**
+	 * Comma-separated list of selectors, no spaces.
+	 *
+	 * @example 'p,div,h1,.css-class,#id'
+	 */
+	selector?: string | undefined;
+	schema?: TransformRawSchema | undefined;
+	isMatch?( node: Node ): boolean;
+	transform?( node: Node ): WPBlock< Partial< T > > | void;
+	[ key: string ]: any; // Permit untyped/undocumented property access.
+}
+
+export interface TransformShortcode< T extends Record< string, any > > {
```

Comment:
It would be lovely to make these types exhaustive and don't allow arbitrary properties.

Reply:
I agree, I can work on that though was warned by @dmsnell that there are concerns about breaking other packages because they might rely on undocumented properties.

Reply:
I can change it, though I was warned it could prevent acceptance and might break other packages.

Reply:
Yeah, finding all the possible properties is challenging. Exhaustive types or not, this PR is valuable anyway.